compiler/IREmitter/Payload/codegen-payload.c

```diff
@@ -2319,33 +2324,45 @@ void sorbet_throwReturn(VALUE retval) {
 // because this function will be inlined, meaning that `tag->buf` will reflect the
 // state of this function's caller and `tag->buf` will be valid for the lifetime of
 // this function's caller.
+//
+// The `volatile` here is a bit gross.  If you look at `rb_ec_tag_state` and its
+// accompanying logic, referenced below, you'll notice that it contains some
+// gross hacks to force clang to stick `ec` in memory before accessing it and that
+// its `ec` argument is `rb_execution_context_t *ec`.
+//
+// Without those hacks, clang will assume that wherever `ec` got allocated to is where
+// it should be accessed on both sides of the `if`, which is not necessarily valid in
+// the case of `longjmp`'ing back to the `if`.  (clang would presumably get this
+// right if the default implementation of RUBY_SETJMP was the C library's `setjmp`,
+// which gets annotated as "returns_twice", but the default implementation is
+// `__builtin_setjmp`, which is not so annotated.)  So the point of the hacks is
+// to force `ec` to (somehow) be allocated to memory.
+//
+// We need a similar hack, but can be slightly more elegant because we cache the
+// execution context across an exception handling region.  A pointer to that cache
+// is passed in here and by making the pointer volatile, we force clang to reload
+// the execution context pointer every time the execution context is needed, thus
+// ensuring that the execution context lives in memory always.
 SORBET_INLINE
-enum ruby_tag_type sorbet_initializeTag(struct rb_vm_tag *tag) {
-    rb_execution_context_t *ec = GET_EC();
-
+enum ruby_tag_type sorbet_initializeTag(volatile rb_execution_context_t **ec, struct rb_vm_tag *tag) {
     // inlined from EC_PUSH_TAG
     tag->state = TAG_NONE;
     tag->tag = Qundef;
-    tag->prev = ec->tag;
+    tag->prev = (*ec)->tag;
 
-    int setjmp_retval = RUBY_SETJMP(tag->buf);
-    enum ruby_tag_type state = TAG_NONE;
-
-    if (setjmp_retval) {
+    if (RUBY_SETJMP(tag->buf) != 0) {
         // See rb_ec_tag_state:
         // https://github.com/ruby/ruby/blob/5445e0435260b449decf2ac16f9d09bae3cafe72/eval_intern.h#L160-L167
-        state = ec->tag->state;
-        ec->tag->state = TAG_NONE;
+        enum ruby_tag_type *statePtr = &(*ec)->tag->state;
+        enum ruby_tag_type state = *statePtr;
+        *statePtr = TAG_NONE;
+        return state;
```

Comment:
Is `statePtr` here just for efficiency's sake to avoid repeating the indirection through `ec` and `ec->tag`, or is `volatile`-ity somehow part of the motivation?